### PR TITLE
Task/module 5

### PR DESCRIPTION
**Pull request by AB619**
1. PINIA is added as a module to the application, Movie list, search + search by setup to work through the store.
2. Movie Detail section is shown from the attached JSON file via store.
3. All app features working and communicate trough the store.

### CODE REVIEW

File: src/components/ActionBar.vue [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    else sortByDate.value = true;
+    movieStore.setSortByParam(option);
+};
+</script>
+
+<template>
+    <div class="panel">
+        <div v-if="movieStore.isMovieDetailOpen" class="panel__moviesCount">Films by {{ movieStore.getMovieById.genres[0] }} genre</div>
+        <div v-else class="panel__moviesCount">{{ moviesCount.valueOf() }} movies found</div>
+        <div v-if="!movieStore.isMovieDetailOpen" class="panel__sort">
```
Comment (hicugi) [inline src/components/ActionBar.vue:21]: Don't forget add space before this line for better reading

File: src/components/ActionBar.vue [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+            <div class="panel__sort__heading">SORT BY</div>
+            <div class="panel__sort__options">
+                <p @click="sortHandler('releaseDate')" v-bind:class="{ 'panel__sort__selected': sortByDate }">RELEASE DATE
+                </p>
```
Comment (hicugi) [inline src/components/ActionBar.vue]: Put the `</p>` after the `DATE`

File: src/components/ActionBar.vue [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+const movieStore = useMovieStore();
+
+const sortByDate = ref(true);
+
+const moviesCount = computed(() => {
+    if (movieStore.list) {
+        return movieStore.list.length;
+    }
+    else return 0;
```
Comment (hicugi) [inline src/components/ActionBar.vue]: You can use direct return with optional chaining:
```js
  return movieStore.list?.length || 0;
```

File: src/components/ActionBar.vue [hunk 4]
```diff
@@ -0,0 +1,83 @@
+<script setup>
+import { ref, computed } from "vue";
+import useMovieStore from "../store/movieStore";
+
+const movieStore = useMovieStore();
+
+const sortByDate = ref(true);
```
Comment (hicugi) [inline src/components/ActionBar.vue]: You don't need extra state, you are already using the `sortByParam` state from the `movieStore` store element.
Just export it and use it here.

File: src/components/ActionBar.vue [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+</script>
+
+<template>
+    <div class="panel">
+        <div v-if="movieStore.isMovieDetailOpen" class="panel__moviesCount">Films by {{ movieStore.getMovieById.genres[0] }} genre</div>
+        <div v-else class="panel__moviesCount">{{ moviesCount.valueOf() }} movies found</div>
+        <div v-if="!movieStore.isMovieDetailOpen" class="panel__sort">
+            <div class="panel__sort__heading">SORT BY</div>
+            <div class="panel__sort__options">
+                <p @click="sortHandler('releaseDate')" v-bind:class="{ 'panel__sort__selected': sortByDate }">RELEASE DATE
```
Comment (hicugi) [inline src/components/ActionBar.vue]: You can't use element in element, look for BEM rules.
You don't have to use the `v-bind` thing:

```html
<p class="{ 'panel__sort--selected': sortByDate }" @click="sortHandler('releaseDate')">RELEASE DATE</p>
```

File: src/components/ActionBar.vue [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                </p>
+                <p @click="sortHandler('imdbRating')" v-bind:class="{ 'panel__sort__selected': !sortByDate }">RATING</p>
+            </div>
+        </div>
+    </div>
+</template>
+
+<style scoped>
```
Comment (hicugi) [inline src/components/ActionBar.vue:31]: The `scoped` not necessary here, if you don't use the `.panel` outside of this component

File: src/components/ActionBar.vue [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+.panel__sort__options>p {
+    color: #ffffff;
+    font-weight: 400;
+    letter-spacing: 0.9px;
+    padding: 5px 15px;
+    background-color: #424242;
+}
+
+.panel__sort__options>.panel__sort__selected {
```
Comment (hicugi) [inline src/components/ActionBar.vue]: Dont need to include the `.panel__sort__options >` extra selector, just `.panel__sort--selected` is enough

File: src/components/Banner.vue [hunk 8]
```diff
@@ -0,0 +1,208 @@
+<script setup>
+import { ref, computed } from "vue";
+import useMovieStore from "../store/movieStore";
+
+const movieStore = useMovieStore();
+
+const searchByTitle = ref(true);
```
Comment (hicugi) [inline src/components/Banner.vue]: Same as with the `sortByDate` state

File: src/components/Banner.vue [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        <div class="banner-info__year">
+          <span>{{ selectedMovie.year }}</span>
+          <span>{{ selectedMovie.duration.slice(2).replace('M', '') }} min</span>
+        </div>
+        <div class="banner-info__storyline">{{ selectedMovie.storyline }}</div>
+      </div>
+    </div>
+    <div v-else class="panel">
+      <div class="searchlabel">FIND YOUR MOVIE</div>
+      <div class="searchbar">
```
Comment (hicugi) [inline src/components/Banner.vue]: Use form for parent, so you can submit by pressing an "Enter" button while focusing on the input field:

```html
      <form class="searchbar" @submit.prevent="clickHandler()">
        <input v-model="query" placeholder="Type and Search" />
        <button type="submit">SEARCH</button>
      </form>
```

File: src/components/MovieList.vue [hunk 10]
```diff
@@ -1,35 +1,29 @@
...
-defineProps<{ movies: Movie[] }>();
 </script>
 
 <template>
   <div class="movies">
-    <div v-if="movies.length > 0" class="movies__list">
-      <div v-for="(m, i) in movies" class="card" :key="i">
-        <img class="card__img" v-lazyload="m.poster" :alt="m.name"/>
+    <div v-if="movies" class="movies__list">
+      <div v-for="(m) in movies" class="card" :key="m.id" @click="movieStore.toggleMovieDetailPanel(m.id)">
```
Comment (hicugi) [inline src/components/MovieList.vue]: Don't include the storage callback here, it's bad approach.
Better to have internal callback in `script`

File: src/components/MovieList.vue [hunk 11]
```diff
@@ -62,13 +56,13 @@ defineProps<{ movies: Movie[] }>();
...
-  max-width: 300px;
+  width: 300px;
   padding: 5px;
 }
 
 .card__img {
-  min-width: 300px; 
-  min-height: 450px;
+  width: 300px; 
+  height: 450px;
```
Comment (hicugi) [inline src/components/MovieList.vue]: Some elements don't have image, here is an example:

image

File: src/components/Banner.vue [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+        <span>roulette</span>
+      </div>
+      <div v-if="movieStore.isMovieDetailOpen" class="header__right">
+        <img src="../assets/search.svg" @click="movieStore.toggleMovieDetailPanel(null)">
+      </div>
+    </header>
+    <div v-if="movieStore.isMovieDetailOpen" class="banner">
+      <img class="banner-img"
+        :src="selectedMovie.posterurl"
+        :placeholder="selectedMovie.title" />
```
Comment (hicugi) [inline src/components/Banner.vue]: The `img` tag doesn't have the `placeholder` attribute: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/img
Use `alt` instead

File: src/store/movieStore.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            if(typeof val === 'number'){
+                setFilterParam("genres");
+            }
+            else{
+                setFilterParam("title");
+                setSearchQuery("");   
+            }
+        }
+
+        const getMovieById = computed(() =>{
```
Comment (hicugi) [inline src/store/movieStore.ts]: The naming is misleading, implying that is method which will return a movie by id.
You can use `selectedMovie`

File: src/store/movieStore.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+            if(isMovieDetailOpen.value){
+                const movie = initialList.value.find((movie: IMockMovie) => movie.id === isMovieDetailOpen.value);
+                movie ? setSearchQuery(movie.genres[0]) : undefined;
```
Comment (hicugi) [inline src/store/movieStore.ts]: Conditional assignment use for returning value or assigning a value.
Use `if` condition for this.

Even better, if you will have a watcher for the `selectedMovie`: https://vuejs.org/guide/essentials/watchers.html

File: src/store/movieStore.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+            }  
+
+            if (sortByParam.value) {
+                filteredList = filteredList.sort((a: IMockMovie, b: IMockMovie) => {
+                    const aValue = a[sortByParam.value];
+                    const bValue = b[sortByParam.value];
+
+                    if (typeof aValue === 'string' && typeof bValue === 'string') {
+                        return aValue.localeCompare(bValue);
+                    } else {
```
Comment (hicugi) [inline src/store/movieStore.ts]: Don't need the `else` statement, if the previous condition returning any value

File: src/store/movieStore.ts [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+                movie ? setSearchQuery(movie.genres[0]) : undefined;
+                return movie;
+            }
+            return undefined;
+        });
+
+        const list = computed(() => {
+            let filteredList: IMockMovie[] = [...initialList.value];
+            if (searchQuery) {
+                searchMovies();
```
Comment (hicugi) [inline src/store/movieStore.ts]: It's quite bad approach to use callback in `computed` function: https://vuejs.org/guide/essentials/computed.html#getters-should-be-side-effect-free

Comment (hicugi) [reply src/store/movieStore.ts]: p.s. don't use chain of events for this kind of stuff

Comment (AB619) [reply src/store/movieStore.ts]: keeping this open to discuss.

**Pull request merged**